### PR TITLE
Auctions#420: live links login

### PR DESCRIPTION
**Pull request by erikdstock**
https://github.com/artsy/auctions/issues/420 tracks this.
This pr standardizes our live auction linking to use a method in a new global `/utils` folder.

The `liveAuctionUrl` method takes one argument for the saleId, appending `/login` if there is a sharify CURRENT_USER.

To test both states, we can look at:
- my active bids components (should only show /login because there is a user)
- auction pages
- artwork pages
- the banners
- loading a bid window before and placing a bid after the sale goes live

When possible, I passed helpers to templates with the locals structure `helpers: {auction: {liveAuctionUrl}}`.

the `Sale` model has many other methods that might want to live in /utils.

### CODE REVIEW

File: desktop/components/my_active_bids/helpers.coffee [hunk 1]
```diff
@@ -1,6 +1,6 @@
 moment = require 'moment'
-{ PREDICTION_URL } = require('sharify').data
+Sale = require '../../models/sale.coffee'
 
 module.exports =
   liveAuctionUrl: (saleId) ->
-    "#{PREDICTION_URL}/#{saleId}"
+    (new Sale(id: saleId)).liveAuctionUrl(true)
```
Comment (sweir27) [inline desktop/components/my_active_bids/helpers.coffee]: In this case, since it's trying to skip the `user` check and this was intentionally avoiding backbone models, I think it's fine to just return `"#{PREDICTION_URL}/#{saleId}/login"` here.

File: desktop/apps/artwork/components/auction/helpers.coffee [hunk 2]
```diff
@@ -1,12 +1,17 @@
 { compact } = require 'underscore'
+Auction = require '../../../../models/auction.coffee'
```
Comment (sweir27) [inline desktop/apps/artwork/components/auction/helpers.coffee]: `Auction` is deprecated (use `Sale` instead).

File: desktop/apps/artwork/components/auction/helpers.coffee [hunk 3]
```diff
@@ -1,12 +1,17 @@
...
 
 pluralize = (word, count, irregular = null) ->
   if count is 1
     "#{count} #{word}"
   else
     "#{count} #{irregular or word + 's'}"
 
+liveAuctionUrl = (auction_id, user) ->
+  (new Auction(id: auction_id)).liveAuctionUrl(user)
```
Comment (sweir27) [inline desktop/apps/artwork/components/auction/helpers.coffee]: As this view is metaphysics-backed, it feels weird to mix in a Backbone model just for the sake of this helper. Makes me wonder if we move the `live_auction_url` logic to metaphysics for use in this view and the `my_active_bids` one.

**Comment by craigspaeth** [comment]
LGTM! I think Sarah makes some good points about the Backbone vs. non-Backbone worlds—something [we've pondered](https://artsy.slack.com/archives/C0350895E/p1488494785987356) before and often just move the logic to Metaphysics in those cases. FWIW I don't think it's necessarily _wrong_ to wrap partial data in Backbone for it's model functionality and therefore wouldn't consider this a blocker—but more and more it becomes a fork the road we should probably make a decision on and stick to.

**Comment by sweir27** [comment]
Ah! Thanks for linking to that slack convo. 😄 I know @damassi had started creating `utils` folders (nested as far as possible) where necessary to accomplish this on the force side. Seems like we could continue that pattern here and keep all of the MP-model-helper logic in one place. /shrug

**Comment by erikdstock** [comment]
@sweir27 I'll come back to this with your recs in mind and then we can merge. Thanks for your comments.

**Comment by damassi** [comment]
I'd still very much like to have a `domain/auctions` bucket somewhere to drop miscellaneous stuff :)

File: desktop/apps/artwork/components/auction/helpers.coffee (hunk 3 above)
Comment (damassi) [inline desktop/apps/artwork/components/auction/helpers.coffee]: I agree -- this seems a little bit unclear, and could also create unwanted, unknown side-effects further down the line should things change. To get a url, one shouldn't have to execute a `constructor`.

File: desktop/apps/artwork/components/auction/view.coffee [hunk 4]
```diff
@@ -7,7 +7,7 @@ openBuyersPremiumModal = require './components/buyers_premium/index.coffee'
 AuthModalView = require '../../../../components/auth_modal/view.coffee'
 inquire = require '../../lib/inquire.coffee'
 acquire = require '../../lib/acquire.coffee'
-helpers = require './helpers.coffee'
+ViewHelpers = require './helpers.coffee'
```
Comment (damassi) [inline desktop/apps/artwork/components/auction/view.coffee]: Why change the name? `helpers` as a var maps directly to the filename, which makes things easier to track down and refactor later

File: desktop/apps/auction/routes.js [hunk 5]
```diff
@@ -19,17 +19,17 @@ export const index = async (req, res) => {
     articles: auctionArticles,
     auction: newAuction,
     footerItems: footerItems,
+    ViewHelpers: { liveAuctionUrl },
```
Comment (damassi) [inline desktop/apps/auction/routes.js]: Unless your using these view helpers as static methods on a class (`ViewHelpers.doSomething()`) or as something instantiatable (`new ViewHelpers()`) I would stick with either `helpers` (following existing convention) or `viewHelpers` (idiomatic js)

File: desktop/apps/auction/templates/banner.jade [hunk 6]
```diff
@@ -8,7 +8,7 @@ if auction.isLiveOpen()
     .auction-banner-live-details
       h1 Live Bidding Now Open
       a.avant-garde-button-white(
-        href="#{sd.PREDICTION_URL}/#{auction.get('id')}#{user ? '/login' : ''}"
+        href=ViewHelpers.liveAuctionUrl(auction.get('id'))
```
Comment (damassi) [inline desktop/apps/auction/templates/banner.jade]: `helpers.liveAuctionUrl()`

File: desktop/apps/artwork/components/auction/view.coffee (hunk 4 above)
Comment (erikdstock) [inline desktop/apps/artwork/components/auction/view.coffee]: I'm open to either. in one component [main_layout i believe] there is some logic around an already existing `helpers`, which i think is why in artwork/auction/templates/live.jade above the `helpers` gets namespaced. Then in my_active_bids there is a ViewHelpers, which I spent a while trying to track down before @sweir27 showed me the way. I intended to be consistent but looks like I still missed some.

File: desktop/components/my_active_bids/helpers.coffee [hunk 7]
```diff
@@ -1,6 +1,6 @@
 moment = require 'moment'
-{ PREDICTION_URL } = require('sharify').data
+Sale = require '../../models/sale.coffee'
+{PREDICTION_URL} = require('sharify').data
+{ liveAuctionUrl } = require '../../../utils/urls.js'
 
-module.exports =
-  liveAuctionUrl: (saleId) ->
-    "#{PREDICTION_URL}/#{saleId}"
+module.exports = liveAuctionUrl
```
Comment (damassi) [inline desktop/components/my_active_bids/helpers.coffee]: Rather than locking down `helpers.coffee` to a single function `module.exports === require('helpers') === liveAuctionUrl` I would expand it as an object so that it carries the assumption that `helpers.coffee` (plural naming) can contain many helpers: 
```coffeescript
module.exports =
  liveAuctionUrl: liveAuctionUrl 
```
which later you can use like
```coffeescript
helpers = require 'helpers'
...
url = helpers.liveAuctionUrl
```
Otherwise, with some slight indirection, 
```coffeescript
liveAuctionUrl = require 'helpers' 
```

File: test/utils/urls_test.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
+const rewire = require('rewire')
+const Urls = rewire('../../utils/urls')
```
Comment (damassi) [inline test/utils/urls_test.js]: Similar to comment up at the top, I would reserve this style of Caps casing to strictly `Class` based objects. Since you'll never instantiate this with `new`, and the capital U in `Urls` seems to imply a class, its somewhat indirect.

File: utils/urls.js [hunk 9]
```diff
@@ -0,0 +1,3 @@
+const { PREDICTION_URL, CURRENT_USER } = require('sharify').data
+
+export const liveAuctionUrl = (id) => `${PREDICTION_URL}/${id}${CURRENT_USER ? '/login' : ''}`
```
Comment (damassi) [inline utils/urls.js]: Thanks for adding this :) Per previous discussions in slack and elsewhere I would update the file location path to be grouped by domain: `utils/domain/auctions/urls.js`.

Comment (damassi) [inline utils/urls.js]: @sweir27 - what do you think?

File: desktop/components/my_active_bids/helpers.coffee (hunk 7 above)
Comment (erikdstock) [inline desktop/components/my_active_bids/helpers.coffee]: Also this file is really not needed so I am going to delete it.

File: utils/urls.js (hunk 9 above)
Comment (erikdstock) [inline utils/urls.js]: even for `urls`?

File: desktop/apps/artwork/components/auction/view.coffee (hunk 4 above)
Comment (damassi) [inline desktop/apps/artwork/components/auction/view.coffee]: Sounds good 👍 Just note that as a rule of thumb, keeping the import var and the filename 1:1 (if not destructuring) is a best practice that will reduce complexity in the long run.

File: utils/urls.js (hunk 9 above)
Comment (sweir27) [inline utils/urls.js]: I think it makes sense to scope by domain! Even just `utils/auctions.js` if we don't want to make too many assumptions about future helper needs.

Comment (damassi) [inline utils/urls.js]: Yup, lets stick to our department :) We shouldn't be worrying about global concerns just yet, and instead opt in to refactoring those out later (if need be) all at once.

File: desktop/apps/auction/routes.js (hunk 5 above)
Comment (erikdstock) [inline desktop/apps/auction/routes.js]: 👍

**Comment by erikdstock** [comment]
Just noticed that CURRENT_USER is not available from sd when utils get loaded. I could go back to my original method of taking a second param forceLogin that could use true, false, or something like `CurrentUser.orNull()`, or we could load that part of the data sooner?

File: utils/domain/auctions/urls.js [hunk 10]
```diff
@@ -0,0 +1,3 @@
+const { PREDICTION_URL, CURRENT_USER } = require('sharify').data
+
+export const liveAuctionUrl = (id) => `${PREDICTION_URL}/${id}${CURRENT_USER ? '/login' : ''}`
```
Comment (damassi) [inline utils/domain/auctions/urls.js]: Regarding your comment, I would suggest using `CurrentUser.orNull()`, but be sure to guard against conditions where it is returned `null`. Also, I would avoid putting logic inside of template tags; something like this is more readable, even if its at the expense of a few more lines of code:

```javascript
export function liveAuctionUrl (id) {
  const currentUser = CurrentUser.orNull()

  if (currentUser) {
    return `${PREDICTION_URL}/${id}/login`
  } else {
    return '/some-fallback-url-so-you-dont-have-to-guard-everywhere-this-function-is-used'
  }
}

File: test/utils/domain/auctions/urls_test.js [hunk 11]
```diff
@@ -0,0 +1,18 @@
+const rewire = require('rewire')
+const urls = rewire('../../../../utils/domain/auctions/urls')
+
+describe('Url Utilities', () => {
+  before(() => {
+    urls.__set__('PREDICTION_URL', 'http://www.liveauctions.com')
+  })
+  describe('liveAuctionUrl()', () => {
```
Comment (damassi) [inline test/utils/domain/auctions/urls_test.js]: Throughout Force people tend to write `describe('#liveAuctionUrl', () => { ... ` to denote functions (hash)

File: utils/domain/auctions/urls.js (hunk 10 above)
Comment (erikdstock) [inline utils/domain/auctions/urls.js]: Unfortunately `CurrentUser.orNull()` is not for use on the server and is basically calling the exact same code.
https://github.com/artsy/force/blob/master/lib/current_user.coffee#L283

Comment (damassi) [inline utils/domain/auctions/urls.js]: Ah yup yup, need to find where that sharify var is being set then

Comment (erikdstock) [inline utils/domain/auctions/urls.js]: I'm just going to move to explicitly pass a user or null in :+1:

File: test/utils/domain/auctions/urls_test.js (hunk 11 above)
Comment (erikdstock) [inline test/utils/domain/auctions/urls_test.js]: 👍

**Comment by erikdstock** [comment]
bid_form.coffee
image

my active bids- user settings screen
image

MAB - auctions list [hard to capture in screenshot]
image


banner - logged in:
image

banner - not logged in: 
image

File: desktop/apps/auction/templates/my_active_bids.jade [hunk 12]
```diff
@@ -23,7 +23,7 @@ if myActiveBids && myActiveBids.length
         .auction-my-active-bids__bids-num (#{bidCount} Bid#{bidCount > 1 ? 's' : ''})
         if bid.sale_artwork.sale.is_live_open
           a.avant-garde-button-white.auction-my-active-bids__bid-live-button(
-              href=ViewHelpers.liveAuctionUrl(bid.sale_artwork.sale_id)
+              href=helpers.auction.liveAuctionUrl(bid.sale_artwork.sale_id, {user: true})
```
Comment (damassi) [inline desktop/apps/auction/templates/my_active_bids.jade]: I wouldn't switch types between calls unless the name is updated as well -- for consistency can you pass in `me` here? or maybe update the var name to be `isLoggedIn`? Then you'd be able to do `isLoggedIn: Boolean(me)`

Comment (erikdstock) [inline desktop/apps/auction/templates/my_active_bids.jade]: 👍

File: desktop/utils/domain/auctions/urls.js [hunk 13]
```diff
@@ -0,0 +1,17 @@
+const sd = require('sharify').data
+
+/** Get the live auction url with /login appended when a user is present
+ * @param {String} id - auction id/slug
+ * @param {Object} options
+ * @param {Boolean} options.isLoggedIn - whether there is a user present
```
Comment (damassi) [inline desktop/utils/domain/auctions/urls.js]: Super small (but good to know) -- with [jsdoc syntax](http://usejsdoc.org/tags-param.html) you wrap optional parameters in brackets. So it would be 

`[options] ... ` 
`[options.isLoggedIn] ...`

File: desktop/components/main_layout/templates/index.jade [hunk 14]
```diff
@@ -1,6 +1,6 @@
 //- Override any locals with `append locals`
 block locals
-  - bodyClass = helpers ? helpers.buildBodyClass(sd, 'body-header-fixed') : '';
+  - bodyClass = helpers && helpers.buildBodyClass ? helpers.buildBodyClass(sd, 'body-header-fixed') : '';
```
Comment (sweir27) [inline desktop/components/main_layout/templates/index.jade]: Why is this change necessary?

**Comment by sweir27** [comment]
This looks good to me! Just had one question (and I think you need to rebase).

File: desktop/components/main_layout/templates/index.jade (hunk 14 above)
Comment (erikdstock) [inline desktop/components/main_layout/templates/index.jade]: `main_layout` gets called from multiple places and assumes that method exists on `helpers`... in some places we pass `helpers` into a template, but in others we use a namespaced `ViewHelpers`. I just chose `helpers`.

File: desktop/apps/auction/test/routes.js [hunk 15]
```diff
@@ -102,7 +102,7 @@ describe('#redirectLive', () => {
     RoutesRewireApi.__Rewire__('metaphysics', sinon.stub().returns(Promise.resolve(auctionQueries)))
     res = {
       redirect: (url) => {
-        url.should.equal('https://live.artsy.net/foo/login')
+        url.should.equal(`${process.env.PREDICTION_URL}/foo/login`)
```
Comment (damassi) [inline desktop/apps/auction/test/routes.js]: I'm not sure `process.env` is available here, and in general it shouldn't be used (these things should all go through `sharify`)

File: desktop/components/main_layout/templates/index.jade (hunk 14 above)
Comment (damassi) [inline desktop/components/main_layout/templates/index.jade]: can do `helpers && helpers.buildBodyClass(sd, 'body-header-fixed') || ''` -- but why would `helpers` not be here?

File: desktop/test/utils/domain/auctions/urls_test.js [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+const urls = rewire('../../../../utils/domain/auctions/urls')
+
+describe('Url Utilities', () => {
+  before(() => {
+    urls.__set__('sd', {PREDICTION_URL: 'http://www.liveauctions.com'})
+  })
+  describe('#liveAuctionUrl builds a url to a live auction', () => {
+    it('appends /login with a true "isLoggedIn" option in the second argument', () => {
+      urls.liveAuctionUrl('foo', { isLoggedIn: Boolean({'currentUserOrNull': 'currentUser'}) }).should.equal('http://www.liveauctions.com/foo/login')
+    })
```
Comment (damassi) [inline desktop/test/utils/domain/auctions/urls_test.js]: Noticed you replaced this above with `PREDICTION_URL`; can you keep this consistent? Also note that its ok to use whatever url as long as your not testing the correctness of the domain.

File: desktop/apps/auction/test/routes.js (hunk 15 above)
Comment (erikdstock) [inline desktop/apps/auction/test/routes.js]: As far as I can tell, sharify isn't available yet in this express handler. I changed the original function to fall back to process.env following the comment here:
https://github.com/erikdstock/force/blob/auctions-420-live-links-login/config.js#L4

File: desktop/test/utils/domain/auctions/urls_test.js [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+    it('does not append login with a falsey user option', () => {
+      urls.liveAuctionUrl('foo', {isLoggedIn: Boolean(null)}).should.equal('http://www.liveauctions.com/foo')
```
Comment (damassi) [inline desktop/test/utils/domain/auctions/urls_test.js]: why not just `isLoggedIn: false`?

File: desktop/utils/domain/auctions/urls.js [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+
+/** Get the live auction url with /login appended when a user is present
+ *  Uses sd.PREDICTION_URL or falls back to process.env in case sd is not fully initialized yet
+ * @param {String} id - auction id/slug
+ * @param {Object} [options]
+ * @param {Boolean} [options.isLoggedIn] - whether there is a user present
+ * @return {String} live auction url
+ */
+export const liveAuctionUrl = (id, options = {}) => {
+  const liveAuctionRoot = sd.PREDICTION_URL || process.env.PREDICTION_URL
```
Comment (damassi) [inline desktop/utils/domain/auctions/urls.js]: I would never rely on `process.env`; it shouldn't be necessary with Sharify and if it is, then there's likely a problem somewhere in the code pipeline.

Comment (erikdstock) [inline desktop/utils/domain/auctions/urls.js]: There are cases where a redirect happens seemingly before the sharify data is loaded. I went to use the config module (linked in a comment above) and it said to use `process.env`... Likewise that is why i use process.env in the routes.js redirect test. I'd be open to doing this another way but I'm just getting conflicting info on which approach to take at this point.

**Comment by erikdstock** [comment]
I think this is good to go but I cannot make its tests pass. I'm out of ideas.

File: desktop/utils/domain/auctions/urls.js [hunk 19]
```diff
@@ -0,0 +1,18 @@
+const sd = require('sharify').data
+const config = require('../../../../config')
+/** Get the live auction url with /login appended when a user is present
+ * @param {String} id - auction id/slug
+ * @param {Object} [options]
+ * @param {Boolean} [options.isLoggedIn] - whether there is a user present
+ * @return {String} live auction url
+ */
+export const liveAuctionUrl = (id, options = {}) => {
+  const liveAuctionRoot = sd.PREDICTION_URL || config.PREDICTION_URL
```
Comment (craigspaeth) [inline desktop/utils/domain/auctions/urls.js]: It appears that you might be using this util client-side which is why tests are [failing on this line](https://github.com/artsy/force/blob/master/desktop/config.coffee#L112), but more importantly you don't want to require config directly—sharify should be enough because it works server & client. So this line should just be `sd.PREDICTION_URL`. Curious why the need for ` || config.`?

File: desktop/utils/domain/auctions/urls.js (hunk 18 above)
Comment (craigspaeth) [inline desktop/utils/domain/auctions/urls.js]: Ah, I see now... that race condition you describe where sharify data isn't loaded before the redirect seems strange. The loading of config data into sharify should happen at "compile time" [around here](https://github.com/artsy/force/blob/master/lib/setup.js#L27). Maybe we can screen hero to debug this problem.

**Comment by erikdstock** [comment]
Ah, @craigspaeth, I was going to ask if you wanted to pair on the failing tests but I was able to resolve them today- the issue with PREDICTION_URL seems to be fixed, or at least I can't reproduce it any more. Thanks for pointing out where things were going wrong in tests. This seems to be working with the correct redirect URLs now. Ready for merge.

File: utils/domain/auctions/urls.js [hunk 20]
```diff
@@ -0,0 +1,19 @@
+const sd = require('sharify').data
+
+/** Get the live auction url with /login appended when a user is present
+ * @param {String} id - auction id/slug
+ * @param {Object} [options]
+ * @param {Boolean} [options.isLoggedIn] - whether there is a user present
+ * @return {String} live auction url
+ */
+export const liveAuctionUrl = (id, options = {}) => {
+  console.log(sd)
```
Comment (damassi) [inline utils/domain/auctions/urls.js]: Can remove `console.log`

Comment (erikdstock) [inline utils/domain/auctions/urls.js]: 👨🏻‍🍳

**Comment by sweir27** [comment]
👍
